Add support for Goodcell 3-phase 22kW EV charger

philippedeswert opened the pull request:
Pre-eliminary support for the Goodcell 3-phase 22kW EV charger.

A few issues remain:
* the two other phases current and voltage are not visible.
* the energy reset does not work (might also not be implemented correctly for all I know on the charger)

make-all commented:
You have multiple unnamed sensors with the same class (for at least voltage and current).  These need names to distinguish them from each other (max is one unnamed sensor of each class, and one unnamed sensor without a class).  Basically each entity needs to be uniquely identified by (entity type)_(name or class)  ie. name is used if available, or class if unnamed.

philippedeswert commented:
> You have multiple unnamed sensors with the same class (for at least voltage and current). These need names to distinguish them from each other (max is one unnamed sensor of each class, and one unnamed sensor without a class). Basically each entity needs to be uniquely identified by (entity type)_(name or class) ie. name is used if available, or class if unnamed.

Ofc, this is so obvious now. Well it was getting late last night. Anyway updated my PR with names as I can now see all different phases. I do not know if it is needed to mark them optional, as that *might* make this also work with the 1-phase models.  It seems there might be something I can do with the constraints too, so current can be adjusted to max 16A for the models not going to 32A.

philippedeswert commented:
Rebased and pushed some small tweaks. 

Could not figure out how to use dps 113 to select dps 114 or 115 for the current setting. (As in dps 114 if dps 113 says Max16A, and dps 115 if it is Max32A), If you can guide me on how to do it I can implement and fix that. Are there other changes needed still otherwise?

make-all commented:
Something like:
```
  - entity: number
    name: Set charge current
    category: config
    icon: "mdi:ev-plug-type2"
    dps:
      - id: 115
        type: integer
        name: value
        unit: A
        range:
          min: 6
          max: 32
        mapping:
          - constraint: max_current
            conditions:
              - dps_val: Max16A
                value_redirect: value_alt
                range:
                  min: 6
                  max: 16
      - id: 114
        type: integer
        name: value_alt
        unit: A
        range:
          min: 6
          max: 16
      - id: 113
        type: string
        name: max_current
 
```

philippedeswert commented:
Updated as per your suggestion, but I am not sure it works (did not seem to switch to pds 114 when I changed the Max16A condition to Max32A to trigger it....

Otherwise it pretty much works only exception is that energy reset button does not seem to work, but it is not supported in the app. So I guess it is not supposed to work. Should I remove that?

Otherwise anything I should do before you consider this done?

make-all commented on custom_components/tuya_local/devices/goodcell_ev_charger.yaml:
```diff
@@ -0,0 +1,275 @@
+        range:
+          min: 6
+          max: 16
+      - id: 113
```
This may need to be a switch or select entity as well if the intention is for the user to make this change.  I am not sure though whether this is an install time option, a user option, or automatic based on vehicle capabilities.

make-all commented on custom_components/tuya_local/devices/goodcell_ev_charger.yaml:
```diff
@@ -0,0 +1,275 @@
+      type: string
+      optional: true
+      name: alarm_set_2
+    - id: 14
```
I think this should be a select entity to select the mode of charging.  Otherwise the settings linked to inactive charging modes will be redundant if you cannot switch to that mode.

make-all commented:
Energy reset probably only affects balance_energy, which is probably only relevant in `charge_energy` charging mode.  I don't think it affects the total_forward_energy, or the charge_energy_once (which I think is the last/current charge only)

make-all commented on custom_components/tuya_local/devices/goodcell_ev_charger.yaml:
```diff
@@ -0,0 +1,275 @@
+        range:
+          min: 6
+          max: 32
+          mapping:
```
This `mapping` should be at the same level of indentation as `range`.  Currently it is under `range`, which prevents this from working.
